### ISSUE TITLE
Module detection from classpath bug

### ISSUE THREAD

**Issue by ascopes** [ascopes/java-compiler-testing#107]
Modules are not being detected from the classpath properly when their sources reside within JARs.

Seems to be related to automatic module handling. Need to work out how Javac handles modules that appear on the classpath within JARs currently.

Reproduction is currently in the Dagger acceptance test.

`src/test/groovy/io/github/ascopes/jct/acceptancetests/dagger/DaggerTest.groovy`
```java
class DaggerTest {
  @DisplayName("Dagger DI runs as expected in the annotation processing phase with modules")
  @JavacCompilers(modules = true)
  @ParameterizedTest(name = "for {0}")
  void daggerDiRunsAsExpectedInTheAnnotationProcessingPhaseWithModules(Compilable compiler) {
    // Given
    def sources = named("sources")
        .copyTreeFrom(
            Path.of("src", "test", "resources", "code"),
            "org/example"
        )
        .createFile(
            "module-info.java",
            """
            |module org.example {
            |  requires dagger;
            |  requires java.annotation;
            |  requires java.base;
            |  exports org.example;
            |  opens org.example to dagger;
            |}
            """.stripMargin()
        )

    // When
    def compilation = compiler
        .addSourcePath(sources)
        .fileManagerLoggingMode(LoggingMode.ENABLED)
        .diagnosticLoggingMode(LoggingMode.STACKTRACES)
        .compile()

    // Then
    assertThatCompilation(compilation).isSuccessfulWithoutWarnings()

    assertThatCompilation(compilation)
        .sourceOutput().packages()
        .withFile("org/example/WebServer_Factory.java").isRegularFile()

    assertThatCompilation(compilation)
        .classOutput().packages()
        .withFile("module-info.class").isRegularFile()

    assertThatCompilation(compilation)
        .classOutput().packages()
        .withFile("org/example/WebServer.class").isRegularFile()

    assertThatCompilation(compilation)
        .classOutput().packages()
        .withFile("org/example/WebServerConfiguration.class").isRegularFile()

    assertThatCompilation(compilation)
        .classOutput().packages()
        .withFile("org/example/WebServer_Factory.class").isRegularFile()
  }
}
```
`src/test/resources/code/WebServer.java`
```java
package org.example;

import javax.inject.Inject;

/**
 * Dummy web server stub.
 */
public class WebServer {
  private final WebServerConfiguration config;

  /**
   * Initialise the web server.
   */
  @Inject
  public WebServer(WebServerConfiguration config) {
    this.config = config;
  }

  public void serve() {
    // do something.
  }
}
```

`src/test/resources/code/WebServerConfiguration.java`
```java
public class WebServerConfiguration {

  final String basePath;
  final Integer port;
  final Boolean https;

  /**
   * Initialise the web server configuration.
   */
  public WebServerConfiguration() {
    basePath = propertyOrEnv("web.server.base.path", Function.identity());
    port = propertyOrEnv("web.server.port", Integer::parseInt);
    https = propertyOrEnv("web.server.https", Boolean::parseBoolean);
  }

  static <T> T propertyOrEnv(String name, Function<String, T> converter) {
    String raw = System.getProperty(name);

    if (raw == null) {
      System.getenv(name.replace('.', '_').toUpperCase(Locale.ROOT));
    }

    if (raw == null) {
      return null;
    }

    return converter.apply(raw);
  }
}
```

Error:

```
java.lang.AssertionError: Expected a successful compilation, but it failed.

Diagnostics:

 - [ERROR] compiler.err.module.not.found /sources/module-info.java (at line 3, col 12)

    1 | 
    2 | module org.example {
    3 |   requires dagger;
      +            ^^^^^^^
    4 |   requires java.annotation;
    5 |   requires java.base;

    module not found: dagger

 - [ERROR] compiler.err.module.not.found /sources/module-info.java (at line 4, col 16)

     2 | module org.example {
     3 |   requires dagger;
     4 |   requires java.annotation;
       +            ^^^^^^^^^^^^^^^^
     5 |   requires java.base;
     6 |   exports org.example;

    module not found: java.annotation
```
